### ISSUE TITLE
NullReferenceException when Projecting Query in EF Core for Cosmos DB

### ISSUE THREAD

**Issue by devbrsa** [dotnet/efcore#31696]
**Description:**

When attempting to project a query result from Cosmos DB using Entity Framework Core, a System.NullReferenceException occurs with the stack trace pointing to lambda_method6.

**Related Issues:**

Issue related from OData: https://github.com/OData/AspNetCoreOData/issues/1027

**Environment:**
- EF Core Package: `Microsoft.EntityFrameworkCore.Cosmos`
- EF Core Version: `6.0.21`
- Database provider: `Microsoft.EntityFrameworkCore.Cosmos`
- Target framework: .NET 6.0
- IDE: (e.g. Visual Studio 2022 17.6)
- Operating System: Windows 10

**Steps to Reproduce:**

```csharp
using Microsoft.EntityFrameworkCore;

var options = new DbContextOptionsBuilder<CosmosDbContext>()
    .UseCosmos("AccountEndpoint=https://localhost:8081/;AccountKey=C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw==", "DummyDb")
    .Options;

using var context = new CosmosDbContext(options);
var employeeList = context.Employee.Select(x => new EmployeeDTO { Id = x.id, Emails = x.User.Emails }).ToList();

public class CosmosDbContext : DbContext
{
    public CosmosDbContext(DbContextOptions<CosmosDbContext> options) : base(options)
    {
    }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        optionsBuilder.UseQueryTrackingBehavior(QueryTrackingBehavior.NoTracking);
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<RootObject>().ToContainer("DummyContainer");
        modelBuilder.Entity<RootObject>().HasNoDiscriminator();
    }

    public DbSet<RootObject> Employee { get; set; }

}

public class RootObject
{
    public string id { get; set; }
    public User User { get; set; }
    public string _rid { get; set; }
    public string _self { get; set; }
    public string _etag { get; set; }
    public string _attachments { get; set; }
    public int _ts { get; set; }
}

public class User
{
    public string Name { get; set; }
    public ICollection<Emails> Emails { get; set; }
}

public class Emails
{
    public string EmailAdress { get; set; }
    public string Type { get; set; }
}

public class EmployeeDTO
{
    public string Id { get; set; }
    public ICollection<Emails> Emails { get; set; }
}
```

Cosmos document representation:

```json
{
    "id": "1",
    "User": {
        "Name": "John",
        "Emails": [
            {
                "EmailAdress": "john@1.com",
                "Type": "A"
            },
            {
                "EmailAdress": "john@2.com",
                "Type": "B"
            }
        ]
    },
    "_rid": "jqYgAN5HGIMBAAAAAAAAAA==",
    "_self": "dbs/jqYgAA==/colls/jqYgAN5HGIM=/docs/jqYgAN5HGIMBAAAAAAAAAA==/",
    "_etag": "\"00000000-0000-0000-e557-09ab7b4901d9\"",
    "_attachments": "attachments/",
    "_ts": 1694508997
}
```

**Expected Behavior:**

I expected to successfully project a query result from Cosmos DB using Entity Framework Core without encountering any exceptions.

**Actual Behavior:**

While attempting to project a query result from Cosmos DB using Entity Framework Core, I encountered a System.NullReferenceException with the following stack trace:
```csharp
System.NullReferenceException: Object reference not set to an instance of an object.
   at lambda_method6(Closure , QueryContext , JObject )
   at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor.QueryingEnumerable`1.Enumerator.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Program.<Main>$(String[] args) in D:\Repos\CosmosBug\CosmosBug\Program.cs:line 8
```

**Comment by devbrsa** [comment]
Configuring the model builder like:

```csharp
    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<RootObject>().ToContainer("DummyContainer");
        modelBuilder.Entity<RootObject>().HasNoDiscriminator();

        modelBuilder.Entity<RootObject>().OwnsOne(r => r.User, user =>
        {
            user.Property(u => u.Name);
            user.OwnsMany(u => u.Emails);
        });
    }
```

The exception is:
```csharp
System.InvalidOperationException: variable 'jObject1' of type 'Newtonsoft.Json.Linq.JObject' referenced from scope '', but it is not defined
   at System.Linq.Expressions.Compiler.VariableBinder.Reference(ParameterExpression node, VariableStorageKind storage)
   at System.Linq.Expressions.Compiler.VariableBinder.VisitParameter(ParameterExpression node)
   at System.Linq.Expressions.ParameterExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitBinary(BinaryExpression node)
   at System.Linq.Expressions.BinaryExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
   at System.Linq.Expressions.Compiler.VariableBinder.VisitBlock(BlockExpression node)
   at System.Linq.Expressions.BlockExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberBinding(MemberBinding node)
   at System.Linq.Expressions.ExpressionVisitor.Visit[T](ReadOnlyCollection`1 nodes, Func`2 elementVisitor)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberInit(MemberInitExpression node)
   at System.Linq.Expressions.MemberInitExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
   at System.Linq.Expressions.Compiler.VariableBinder.VisitLambda[T](Expression`1 node)
   at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.Compiler.LambdaCompiler.Compile(LambdaExpression lambda)
   at System.Linq.Expressions.LambdaExpression.Compile()
   at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor.VisitShapedQuery(ShapedQueryExpression shapedQueryExpression)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at System.Linq.Expressions.Expression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass9_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Program.<Main>$(String[] args) in CosmosBug\Program.cs:line 8
```

**Comment by ajcvickers** [comment]
@devbrsa Please test using [EF Core 8 RC 1](https://www.nuget.org/packages/Microsoft.EntityFrameworkCore/8.0.0-rc.1.23419.6) and report back if you are still seeing an issue.

**Comment by devbrsa** [comment]
> @devbrsa Please test using [EF Core 8 RC 1](https://www.nuget.org/packages/Microsoft.EntityFrameworkCore/8.0.0-rc.1.23419.6) and report back if you are still seeing an issue.

@ajcvickers The error persists in the given version provided (EF Core 8 RC 1 (rc.1.23419.6)).

```csharp
System.NullReferenceException: Object reference not set to an instance of an object.
   at lambda_method6(Closure, QueryContext, JObject)
   at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor.QueryingEnumerable`1.Enumerator.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Program.<Main>$(String[] args)
```

**Comment by devbrsa** [comment]
I've tested using the release for .NET 9.0.100-preview.7.24407.12 and EF 9.0.0-preview.7.24405.3, and the error persists.

**Comment by roji** [comment]
@devbrsa can you please check this with the latest [daily build](https://github.com/dotnet/efcore/blob/main/docs/DailyBuilds.md)? It's possible that the fix wasn't yet merged for preview.7.

**Comment by devbrsa** [comment]
Hi @roji thanks for your quick reply.
Unfortunately same behavior following the daily build version.

To accommodate the code compliance I have changed the call to async:

```csharp
var emailsList = await context.Employee.Select(x => x.User.Emails).ToListAsync();
```

Latest versions used: 
```xml
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="9.0.0-rc.2.24459.8" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Cosmos" Version="9.0.0-rc.2.24459.8" />
```


![image](https://github.com/user-attachments/assets/48a5d78c-09ac-4998-ab2a-a3f8a1c30d7d)

At this stage I'm not sure the feature to project from db side is provisioned once working with Cosmos Db.

**Comment by roji** [comment]
Confirmed - with the latest rc.2 build (9.0.0-rc.2.24460.3) the error reproes with the below code: subfragments don't get projected out correctly when projecting out an anonymous type:

```c#
var employee = await context.Employee
    .Select(x => new EmployeeDTO { Id = x.id, Name = x.User.Name, Emails = x.User.Emails })
    .SingleAsync();
Console.WriteLine($"Name null: {employee.Name is null}, Emails null: {employee.Emails is null}"); // email is null though it isn't in the database
```

Note that the SQL we generate is:

```sql
SELECT c["id"], c["User"]["Name"], c
FROM root c
OFFSET 0 LIMIT 2
```

Note that the emails are projected via `c`, so the entire document is projected though only a part of it is actually needed (this is #34067). We should take a general look at projecting owned types out of the database etc., there seems to be lots of work needed there.

<details>
<summary>Minimal repro</summary>

```c#
await using var context = new CosmosDbContext();
await context.Database.EnsureDeletedAsync();
await context.Database.EnsureCreatedAsync();

context.Employee.Add(new()
{
    id = "1",
    User = new()
    {
        Name = "John",
        Emails = [
            new() { EmailAdress = "john@1.com", Type = "A"},
            new() { EmailAdress = "john@2.com", Type = "B"}
        ]
    }
});
await context.SaveChangesAsync();

context.ChangeTracker.Clear();

var employee = await context.Employee
    .Select(x => new EmployeeDTO { Id = x.id, Name = x.User.Name, Emails = x.User.Emails })
    .SingleAsync();
Console.WriteLine($"Name null: {employee.Name is null}, Emails null: {employee.Emails is null}");

public class CosmosDbContext : DbContext
{
    public DbSet<RootObject> Employee { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .UseCosmos(
                "https://192.168.64.6:8081",
                "C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw==",
                "Test",
                o => o.HttpClientFactory(() => new HttpClient(
                        new HttpClientHandler
                        {
                            ServerCertificateCustomValidationCallback =
                                HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
                        }))
                    .ConnectionMode(ConnectionMode.Gateway))
            .UseQueryTrackingBehavior(QueryTrackingBehavior.NoTracking)
            .LogTo(Console.WriteLine, LogLevel.Information)
            .EnableSensitiveDataLogging();

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<RootObject>().ToContainer("DummyContainer");
        modelBuilder.Entity<RootObject>().HasNoDiscriminator();
    }
}

public class RootObject
{
    public string id { get; set; }
    public User User { get; set; }
    public string _rid { get; set; }
    public string _self { get; set; }
    public string _etag { get; set; }
    public string _attachments { get; set; }
    public int _ts { get; set; }
}

public class User
{
    public string Name { get; set; }
    public ICollection<Emails> Emails { get; set; }
}

public class Emails
{
    public string EmailAdress { get; set; }
    public string Type { get; set; }
}

public class EmployeeDTO
{
    public string Id { get; set; }
    public string Name { get; set; }
    public ICollection<Emails> Emails { get; set; }
}
```

</details>

**Comment by JoasE** [comment]
Might be (partly?) fixed by #34567 once completed
